library(bambu)

# format counts the way SQANTI likes
counts2SQ <- function(counts){
  original_cols <- colnames(counts)
  if (ncol(counts) > 1){
    counts$superPBID <- rownames(counts)
    counts <- counts[, c("superPBID", original_cols)]
  } else {
    counts$pbid <- rownames(counts)
    counts <- counts[, c("pbid", original_cols)]
  }
  return(counts)
}


args <- commandArgs(trailingOnly = TRUE)

genome = args[[1]]
annotaion = args[[2]]
reads = args[[3]]
out_dir = args[[4]]
n_cores = args[[5]]

# Prepare the path to write the outputs
path <- dirname(out_dir)
prefix <- basename(out_dir)

# get the read files
reads <- read.delim(reads, header=F,col.names = c("Cond", "Sample", "Pool", "bam", "fastq"))

reads <- reads$bam

# Load the annotation generated by J&C
annotation <- prepareAnnotations(annotaion)

# Run bambu quantification
se <- bambu(reads = reads, annotations = annotation, genome = genome, discovery = FALSE, ncore = n_cores)

# Get counts in SQANTI3 format
counts <- data.frame(assays(se)$counts)
counts <- counts2SQ(counts)

# Write trasncripts
write.table(counts, 
            file = paste0(path, "/",prefix, "_counts.tsv"), 
            quote = F,
            row.names = F,
            sep = "\t")