#!/bin/bash
#SBATCH --job-name=bambu_quant
#SBATCH --qos short
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=16
#SBATCH --mail-type=ALL # Mail events (NONE, BEGIN, END, FAIL, ALL)
#SBATCH --mem=30gb # Per processor memory
#SBATCH -t 12:00:00     # Walltime
#SBATCH -o log/bambu_quant_%A_%a.out

module load anaconda
source activate bambu_env
# Input
genome="$1"
metadata_concat="$2"
metadata_samples="$3"
out_iso_detect="$4"
out_quant="$5"

line_number_metadata=$((SLURM_ARRAY_TASK_ID + 1)) # Add 1 because of header
n_cores="$SLURM_CPUS_PER_TASK"

# Change working directory
# cd $WD

# Get sample metadata
read -r cond sample_id pool bam_file fastq_file <<< "$(head -n $line_number_metadata $metadata_concat | tail -n 1)"

out_quant="${out_quant}/${cond}/${sample_id}/${sample_id}"
out_bambu="${out_iso_detect}/bambu/${cond}/${sample_id}/${sample_id}"

# Use the annotation generated by the J&C method
annotation=${out_bambu}.gtf

# Print Input
echo "Condition: $cond"
echo "Genome: $genome"
echo "Annotation: $annotation"
echo "Out dir bambu: $out_bambu"
echo "Out dir quant: $out_quant"
echo "N cores: $n_cores"
echo ""

# Check if output directories exist
directories=("$out_iso_detect" $(dirname "$out_quant")  $(dirname "$out_bambu"))

for directory in "${directories[@]}"; do
	if [ ! -d "$directory" ]; then
		mkdir -p "$directory"
		echo "Directory created: $directory"
	else
		echo "Directory already exists: $directory"
	fi
done


# Get the info of the reads associated to the condition
grep $cond $metadata_samples > ${out_quant}_reads.fofn
reads_file=${out_quant}_reads.fofn

# Run bambu quantification
Rscript $SCRIPT_DIR/bambu_quant.R $genome $annotation $reads_file $out_quant $n_cores